src/lib/voiceSearchHelper/index.ts

```diff
@@ -0,0 +1,107 @@
+const STATUS_INITIAL = 'initial';
+const STATUS_ASKING_PERMISSION = 'askingPermission';
+const STATUS_WAITING = 'waiting';
+const STATUS_RECOGNIZING = 'recognizing';
+const STATUS_FINISHED = 'finished';
+const STATUS_ERROR = 'error';
+
+type VoiceSearchHelperParams = {
+  onQueryChange: (query: string) => void;
+  onStateChange: () => void;
+};
+
+export default function voiceSearchHelper({
+  onQueryChange,
+  onStateChange,
+}: VoiceSearchHelperParams) {
+  const SpeechRecognition =
+    (window as any).webkitSpeechRecognition ||
+    (window as any).SpeechRecognition;
+  const getDefaultState = (status: string) => ({
+    status,
+    transcript: '',
+    isSpeechFinal: undefined,
+    errorCode: undefined,
+  });
+  let state = getDefaultState(STATUS_INITIAL);
+  let recognition: any;
+
+  const isSupportedBrowser = () => Boolean(SpeechRecognition);
```

Comment:
This can be a boolean instead of a function. A more correct name could be `isBrowserSupported`.

Reply:
@francoischalifour Right, this can be a boolean. Is `isBrowserSupported` okay to be a boolean while `isListening` is a function?

Reply:
That's a good point. It would make sense to keep the same type for both of them.

Reply:
Okay, then I'll leave both of them as function.